### PR TITLE
Add support for setpriority and getpriority

### PR DESCRIPTION
**Pull request by liqinggd**

### CODE REVIEW

File: services/libs/aster-std/src/syscall/set_get_priority.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                Self::User(uid)
+            }
+        })
+    }
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Clone, Debug, TryFromInt)]
+#[repr(i32)]
+enum c_prio_which {
```
Comment (tatetian) [inline services/libs/aster-std/src/syscall/set_get_priority.rs:120]: Sometimes we choose lower-case names to be compatible with the C definition. But there is not a C type called `c_prio_which`. So in this specific case, using the lower-case name is not wise. I suggest calling it `Which`. It is ok to give a short name as it is an internal type only used in this module. So the short name won't cause any ambiguity.

File: services/libs/aster-std/src/sched/nice.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    raw: i32,
+}
+
+impl Nice {
+    const MIN: i32 = -20;
+
+    const MAX: i32 = 19;
+
+    /// Returns the highest priority.
+    pub fn highest_priority() -> Self {
```
Comment (tatetian) [inline services/libs/aster-std/src/sched/nice.rs:21]: I think having the `highest_priority` and `lowest_priority` methods are redundant with the `MAX` and `MIN`. From the user side, typing `Nice::MAX` is faster than `Nice::highest_priority()`. The same argument can be made to `RlimitNice`.

```rust
impl Nice {
    /// The maximum nice, whose value is 19.
    pub const MAX: Self = Nice { raw: 19 };
}
```

File: services/libs/aster-std/src/process/process/process_group.rs [hunk 3]
```diff
@@ -69,6 +69,11 @@ impl ProcessGroup {
         self.pgid
     }
 
+    /// Returns all processes in the group.
+    pub fn get_all_processes(&self) -> Vec<Arc<Process>> {
+        self.inner.lock().processes.values().cloned().collect()
```
Comment (tatetian) [inline services/libs/aster-std/src/process/process/process_group.rs:74]: When the operation on each process is light-weight enough (like setting a nice value), dumping all processes into a `Vec` seems too costly. I think a more desirable approach is adding an `iter` method.

```rust
pub struct Iter<'a> {
    inner: MutexGuard<'a, ProcessGroupInner>
}

/// Returns an iterator for the processes within the group.
pub fn iter(&self) -> Iter<'a> {
    ...
}
```


This way the user can enjoy the flexibility of choosing between two approaches:
1. Holding the internal lock while iterating the member processes. No copying or creating `Vec`.
2. Calling the `collect` method on the iterator to return a `Vec`.

I recommend adding the `iter` method to process table, process group, and process.

File: services/libs/aster-std/src/process/process/mod.rs [hunk 4]
```diff
@@ -71,6 +72,9 @@ pub struct Process {
     umask: Arc<RwLock<FileCreationMask>>,
     /// resource limits
     resource_limits: Mutex<ResourceLimits>,
+    /// Scheduling priority
+    /// According to POSIX, the nice value is a per-process setting.
+    nice: Mutex<Nice>,
```
Comment (tatetian) [inline services/libs/aster-std/src/process/process/mod.rs:77]: Using `Mutex` looks to too heavy-weight. I can't think of any use cases where the nice value must be accessed within a critical section. How about wrapping an `AtomicU8` inside `Nice`?

File: services/libs/aster-std/src/syscall/set_get_priority.rs (hunk 1 above)
Comment (liqinggd) [reply services/libs/aster-std/src/syscall/set_get_priority.rs]: OK

File: services/libs/aster-std/src/process/process/process_group.rs (hunk 3 above)
Comment (liqinggd) [reply services/libs/aster-std/src/process/process/process_group.rs]: In the beginning, I just followed the method of `get_all_processes` in process_table, the `iter` method is  better.

Done with adding the `iter()` method.

File: services/libs/aster-std/src/process/process/mod.rs (hunk 4 above)
Comment (liqinggd) [reply services/libs/aster-std/src/process/process/mod.rs]: Yes, using `Mutex` seems too heavy, I changed it to AtomicU8 inside `Nice`.

File: services/libs/aster-std/src/sched/nice.rs (hunk 2 above)
Comment (liqinggd) [reply services/libs/aster-std/src/sched/nice.rs]: Done

File: services/libs/aster-std/src/sched/nice.rs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+///
+/// It is a value in the range -20 to 19, with -20 being the highest priority
+/// and 19 being the lowest priority. The smaller values give a process a higher
+/// scheduling priority.
+pub struct Nice {
+    // Use `u8` with the range [0, 39] to represent the real range [-20, 19]
+    value: AtomicU8,
+}
+
+impl Nice {
```
Comment (tatetian) [inline services/libs/aster-std/src/sched/nice.rs:17]: Use `Atomic<Nice>`

File: services/libs/aster-std/src/sched/nice.rs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    }
+}
+
+/// The process scheduling priority value used in rlimit.
+///
+/// It is a value in the range 40 to 1, with 40 being the highest priority
+/// and 1 being the lowest priority. The greater values give a process a higher
+/// scheduling priority.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RlimitNice {
```
Comment (tatetian) [inline services/libs/aster-std/src/sched/nice.rs:119]: Call this `Priority` and explain its relationship with `Nice`.

File: services/libs/aster-std/src/process/process/process_group.rs [hunk 7]
```diff
@@ -90,3 +98,21 @@ impl ProcessGroup {
...
+pub struct ProcessGroupIter<'a> {
+    inner: MutexGuard<'a, Inner>,
+    idx: usize,
+}
+
+impl<'a> Iterator for ProcessGroupIter<'a> {
+    type Item = Arc<Process>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let process = self.inner.processes.values().nth(self.idx);
```
[the remark this answers is not in the document]
Comment (liqinggd) [reply services/libs/aster-std/src/process/process/process_group.rs]: Done

File: services/libs/aster-std/src/sched/nice.rs [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+use core::sync::atomic::{AtomicU8, Ordering};
+
+/// The process scheduling priority value.
+///
+/// It is a value in the range -20 to 19, with -20 being the highest priority
+/// and 19 being the lowest priority. The smaller values give a process a higher
+/// scheduling priority.
+pub struct Nice {
+    // Use `u8` with the range [0, 39] to represent the real range [-20, 19]
+    value: AtomicU8,
```
[the remark this answers is not in the document]
Comment (liqinggd) [reply services/libs/aster-std/src/sched/nice.rs]: Done

File: services/libs/aster-std/src/sched/nice.rs (hunk 5 above)
Comment (liqinggd) [reply services/libs/aster-std/src/sched/nice.rs]: Done

File: services/libs/aster-std/src/sched/nice.rs (hunk 6 above)
Comment (liqinggd) [reply services/libs/aster-std/src/sched/nice.rs]: Done

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**